distribution/push_v2.go

```diff
@@ -470,27 +470,28 @@ func (pd *pushDescriptor) uploadUsingSession(
 		return distribution.Descriptor{}, retryOnError(err)
 	}
 
-	reader = progress.NewProgressReader(ioutils.NewCancelReadCloser(ctx, contentReader), progressOutput, pd.layer.Size(), pd.ID(), "Pushing")
+	readerCtx, cancelReader := context.WithCancel(ctx)
+	reader = progress.NewProgressReader(ioutils.NewCancelReadCloser(readerCtx, contentReader), progressOutput, pd.layer.Size(), pd.ID(), "Pushing")
 
 	switch m := pd.layer.MediaType(); m {
 	case schema2.MediaTypeUncompressedLayer:
 		compressedReader, compressionDone := compress(reader)
-		defer func(closer io.Closer) {
-			closer.Close()
+		defer func() {
+			cancelReader()
```

Comment:
Yes, it's safe:
> After the first call, subsequent calls to a CancelFunc do nothing. 
https://pkg.go.dev/context#CancelFunc


> (wondering if a global defer would work for this code; similar to the pull variant 🤔)

Global `defer` would be more awkward IMO, because we want to do the close before the `<-compressionDone`  and also close it as soon as possible ([after the `ReadFrom` call](https://github.com/moby/moby/blob/2f76dfd473ebcec66954890140244fddd43822b8/distribution/push_v2.go#L493))

Reply:
> we want to do the close before the <-compressionDone

Doh! I was looking at that and still commented. Yup yup, all good